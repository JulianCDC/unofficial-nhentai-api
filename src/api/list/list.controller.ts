import {
  CacheInterceptor,
  CacheTTL,
  Controller,
  Get,
  NotFoundException,
  Query,
  Res,
  UseInterceptors,
} from '@nestjs/common';
import {
  ApiBadRequestResponse,
  ApiNotFoundResponse,
  ApiOperation,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';
import { Response } from 'express';
import { ListService } from './list.service';
import { GetHomepageDto } from './types/get-homepage.dto';
import { HomepageEntity } from './types/homepage.entity';
import { PostListEntity } from './types/post-list.entity';
import { SearchPostDto } from './types/search-post.dto';
import { SearchRandomPostDto } from './types/search-random-post.dto';

@ApiTags('list')
@Controller('list')
@UseInterceptors(CacheInterceptor)
export class ListController {
  constructor(private readonly listService: ListService) {}

  @ApiOperation({
    summary: 'use nhentai search feature ',
    description:
      'for advanced use see [https://nhentai.net/info/](https://nhentai.net/info/)',
  })
  @ApiBadRequestResponse()
  @Get('/search')
  async searchGalleries(
    @Query() query: SearchPostDto,
  ): Promise<PostListEntity> {
    const uri = `https://nhentai.net/search/?q=${query.q}&sort=${query.sort}&page=${query.page}`;

    const [posts, pages] = await this.listService.fetchPosts(uri, query.page);

    return {
      data: posts,
      pagination: {
        page: query.page,
        total: pages,
      },
    };
  }

  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'no gallery found' })
  @ApiResponse({
    status: 303,
    description: 'redirect to random post on nhentai',
  })
  @CacheTTL(1)
  @ApiOperation({
    summary: 'get a random post corresponding to a search query',
  })
  @Get('/randomsearch')
  async searchRandomGallery(
    @Res() res: Response,
    @Query() query: SearchRandomPostDto,
  ): Promise<void> {
    const [, numberOfPages] = await this.listService.fetchPosts(
      `https://nhentai.net/search/?q=${query.q}`,
      1,
    );

    const randomPage = Math.ceil(Math.random() * numberOfPages);

    const [randomPostList] = await this.listService.fetchPosts(
      `https://nhentai.net/search/?q=${query.q}&page=${randomPage}`,
      randomPage,
    );

    const randomPostIndex = Math.round(
      Math.random() * (randomPostList.length - 1),
    );
    const randomPost = randomPostList[randomPostIndex];

    if (!randomPost) {
      throw new NotFoundException();
    }

    if (query.redirect) {
      res.setHeader('Cache-Control', 'no-cache');

      res.redirect(303, `https://nhentai.net/g/${randomPost.id}`);
    } else {
      res.json(randomPost);
    }
  }

  @ApiOperation({ summary: 'get posts in nhentai homepage ' })
  @Get('/homepage')
  async getHomepage(@Query() query: GetHomepageDto): Promise<HomepageEntity> {
    const uri = `https://nhentai.net/?page=${query.page}`;

    const [posts, pages] = await this.listService.fetchPosts(uri, query.page);

    const popular = query.page === 1 ? posts.slice(0, 5) : [];
    const recent = query.page === 1 ? posts.slice(5) : posts;

    return {
      data: {
        popular,
        recent,
      },
      pagination: {
        page: query.page,
        total: pages,
      },
    };
  }
}
