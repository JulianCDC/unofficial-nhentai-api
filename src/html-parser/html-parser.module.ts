import { Module, Global } from '@nestjs/common';
import { HtmlParserService } from './html-parser/html-parser.service';

@Global()
@Module({
  providers: [HtmlParserService],
  exports: [HtmlParserService],
})
export class HtmlParserModule {}
