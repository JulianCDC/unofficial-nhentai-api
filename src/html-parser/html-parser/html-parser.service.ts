import { Injectable } from '@nestjs/common';
import fetch from 'node-fetch';
import { load } from 'cheerio';

@Injectable()
export class HtmlParserService {
  async parse(url: string): Promise<CheerioStatic> {
    const html = await this.fetchPage(url);
    const $ = load(html);

    return $;
  }

  async mapParse<T>(
    url: string,
    selector: CheerioSelector | string,
    map: (element: Cheerio) => T,
  ): Promise<T[]> {
    const $ = await this.parse(url);

    return $(selector)
      .toArray()
      .map(element => map($(element)));
  }

  private async fetchPage(url: string): Promise<string> {
    return fetch(url).then(response => response.text());
  }
}
