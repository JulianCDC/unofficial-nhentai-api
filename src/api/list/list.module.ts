import { CacheModule, Module } from '@nestjs/common';
import { HtmlParserModule } from '../../html-parser/html-parser.module';
import { ListController } from './list.controller';
import { ListService } from './list.service';

@Module({
  imports: [HtmlParserModule, CacheModule.register({ ttl: 3600 })],
  controllers: [ListController],
  providers: [ListService],
})
export class ListModule {}
