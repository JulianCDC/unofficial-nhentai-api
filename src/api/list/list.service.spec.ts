import { CacheModule } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { load } from 'cheerio';
import * as fs from 'fs';
import { HtmlParserModule } from 'src/html-parser/html-parser.module';
import { HtmlParserService } from 'src/html-parser/html-parser/html-parser.service';
import { ListService } from './list.service';

describe('ListService', () => {
  let listService: ListService;
  let htmlParserService: HtmlParserService;

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      controllers: [],
      providers: [ListService],
      imports: [CacheModule.register({ ttl: 1 }), HtmlParserModule],
    }).compile();

    listService = moduleRef.get(ListService);
    htmlParserService = moduleRef.get(HtmlParserService);
  });

  describe('fetchPosts', () => {
    it('Returns posts in homepage', async () => {
      htmlParserService.parse = jest
        .fn()
        .mockReturnValueOnce(
          load(fs.readFileSync('test/mocks/pages/homepage.html')),
        );

      await expect(
        listService.fetchPosts('https://nhentai.net/', 1),
      ).resolves.toMatchSnapshot();
    });
  });
});
