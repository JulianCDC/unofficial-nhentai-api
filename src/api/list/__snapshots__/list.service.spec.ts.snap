// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListService fetchPosts Returns posts in homepage 1`] = `
Array [
  Array [
    Object {
      "id": 326778,
      "lang": "EN",
      "name": "[Mifune Seijirou] Elf Harem Academia Nangoku Kotsukuri Gasshuku Hen (COMIC MILF 2019-12 Vol. 51) [English] {Hennojin} [Digital]",
      "tags": Array [
        12227,
        17249,
        5010,
        35763,
        13989,
        2937,
        19175,
        29859,
        24529,
        19018,
        832,
        93324,
        3666,
        8378,
        105833,
        15785,
        29224,
        24102,
        25871,
        1207,
        13720,
        85288,
        6343,
        24201,
        1590,
        3735,
        28550,
        27112,
        85295,
        20035,
        8368,
        8010,
        21572,
        33173,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720483/thumb.jpg",
    },
    Object {
      "id": 326785,
      "lang": "EN",
      "name": "(C90) [70 Nenshiki Yuukyuu Kikan (Ohagi-san)] The crucifixion of Succura (Bloodborne) [English] [ancilf]",
      "tags": Array [
        12227,
        17249,
        51110,
        426,
        70497,
        9661,
        24555,
        16228,
        832,
        25871,
        21450,
        13720,
        7546,
        20905,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720183/thumb.jpg",
    },
    Object {
      "id": 326779,
      "lang": "EN",
      "name": "[Momo no Suidousui] Saikyou Futanari Tenkousei | The Strongest Futanari Transfer Student (Futanari Friends! 07) [English] {Hennojin}",
      "tags": Array [
        12227,
        17249,
        52015,
        14283,
        19561,
        30555,
        29859,
        35970,
        72471,
        22945,
        779,
        23967,
        29224,
        27384,
        13720,
        10314,
        24201,
        32484,
        28550,
        85295,
        6817,
        33173,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720484/thumb.jpg",
    },
    Object {
      "id": 326638,
      "lang": "EN",
      "name": "(C97) [clesta (Cle Masahiro)] CL-orc 02 PAKO Friend | Sex Friend [English] [Decensored] [Learn JP with H]",
      "tags": Array [
        12227,
        17249,
        90671,
        15564,
        10769,
        8378,
        7752,
        35763,
        2515,
        2937,
        20282,
        25050,
        13720,
        85288,
        10314,
        35762,
        20905,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1718557/thumb.jpg",
    },
    Object {
      "id": 326639,
      "lang": "EN",
      "name": "[DISTANCE] Jyoshi Luck! ~2 Years Later~ Satsuki-sensei no Ninkai Nikki Hen (COMIC ExE 20) [English] [Fated Circle] [Digital]",
      "tags": Array [
        12227,
        17249,
        3518,
        35763,
        81707,
        2937,
        9260,
        24764,
        29224,
        9162,
        1207,
        35762,
        33173,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1719883/thumb.jpg",
    },
    Object {
      "id": 326879,
      "lang": "CN",
      "name": "[Tagame Gengoroh] Gedo no Ie Gekan [Chinese][同文城]",
      "tags": Array [
        29963,
        17249,
        30995,
        14283,
        36957,
        29859,
        15658,
        28349,
        29013,
        19899,
        107705,
        6898,
        22942,
        1067,
        17773,
        9162,
        30473,
        2956,
        2452,
        7752,
        32752,
        33129,
        10542,
        20284,
        23895,
        7142,
        1207,
        8010,
        7256,
        15492,
        8739,
        23237,
        33173,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1721002/thumb.jpg",
    },
    Object {
      "id": 326878,
      "lang": "CN",
      "name": "[Tagame Gengoroh] Gedou no Ie Chuukan [Chinese][同文城]",
      "tags": Array [
        29963,
        17249,
        30995,
        14283,
        73750,
        31880,
        36957,
        29859,
        15658,
        28349,
        3455,
        9260,
        31044,
        29013,
        107705,
        6898,
        4435,
        16828,
        5357,
        22942,
        1067,
        17773,
        21712,
        9162,
        30473,
        5820,
        78262,
        32752,
        33129,
        5529,
        10476,
        20284,
        23895,
        7142,
        1207,
        8010,
        7256,
        15492,
        8739,
        23237,
        33173,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1721001/thumb.jpg",
    },
    Object {
      "id": 326877,
      "lang": "CN",
      "name": "[Tagame Gengoroh] Gedou no Ie Joukan [Chinese][同文城]",
      "tags": Array [
        29963,
        17249,
        30995,
        14283,
        33032,
        73750,
        31012,
        29859,
        15658,
        28349,
        9260,
        29013,
        107705,
        6898,
        4435,
        16828,
        1033,
        5357,
        22942,
        1067,
        17773,
        101724,
        30473,
        13720,
        27553,
        2820,
        14971,
        32752,
        23895,
        25296,
        11020,
        80718,
        15492,
        8739,
        23237,
        33173,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720993/thumb.jpg",
    },
    Object {
      "id": 326876,
      "lang": "CN",
      "name": "[Meisho Hanten (Haruaki)] Juuyoku UNGOGGLED [Chinese] [Digital][同文城]",
      "tags": Array [
        29963,
        17249,
        90671,
        55414,
        55415,
        14283,
        29859,
        19018,
        5620,
        22945,
        107705,
        8010,
        21712,
        30473,
        28550,
        23895,
        8368,
        15492,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720991/thumb.jpg",
    },
    Object {
      "id": 326875,
      "lang": "CN",
      "name": "[Meisho Hanten (Haruaki)] Kamisama to Kawataredoki [Chinese] [Digital][同文城]",
      "tags": Array [
        29963,
        17249,
        90671,
        55414,
        55415,
        14283,
        30555,
        3843,
        1033,
        21712,
        9162,
        30473,
        22950,
        23895,
        8368,
        15492,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720990/thumb.jpg",
    },
    Object {
      "id": 326874,
      "lang": "CN",
      "name": "[Yukemuri Froster (Yunosuke)] Cool Shine Bitch!! encore [Chinese] [Digital][同文城]",
      "tags": Array [
        29963,
        17249,
        90671,
        36951,
        36952,
        14283,
        19018,
        29013,
        107705,
        8378,
        21712,
        30473,
        23895,
        15492,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720989/thumb.jpg",
    },
    Object {
      "id": 326873,
      "lang": "EN",
      "name": "[Dynamite☆Honey] Kunoichi Dynamite DL Ban (Various) [Digital] [English] [EroGPx] [Incomplete]",
      "tags": Array [
        12227,
        17249,
        13159,
        14499,
        23480,
        29013,
        107705,
        14283,
        2937,
        29859,
        22945,
        16828,
        1207,
        13720,
        27553,
        14971,
        8010,
        7256,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720972/thumb.jpg",
    },
    Object {
      "id": 326872,
      "lang": "JP",
      "name": "[Nanairo Otogizoushi (Miyase Mahiro)] SANA maidcafe + SANA maidcafe -Another side- (Touhou Project) [Digital]",
      "tags": Array [
        18024,
        4675,
        9005,
        8999,
        27378,
        20905,
        33172,
        6346,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720948/thumb.jpg",
    },
    Object {
      "id": 326871,
      "lang": "CN",
      "name": "(C95) [Dokomademo Aoi Sora ni Ukabu Niku. (Nikusoukyuu.)] 9a-91-chan wa Miraretai. (Girls' Frontline) [Chinese] [v.v.t.m汉化组]",
      "tags": Array [
        29963,
        17249,
        74788,
        99469,
        92923,
        20572,
        8530,
        35763,
        13989,
        105833,
        2633,
        13720,
        35762,
        24201,
        6817,
        27473,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720942/thumb.jpg",
    },
    Object {
      "id": 326870,
      "lang": "CN",
      "name": "(C89) [Kamishiki (Kamizuki Shiki)] 3-nen K-gumi T-toku Sensei! (Kantai Collection -KanColle-) [Chinese] [白杨汉化组]",
      "tags": Array [
        29963,
        17249,
        1841,
        28040,
        14381,
        28550,
        10314,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720939/thumb.jpg",
    },
    Object {
      "id": 326869,
      "lang": "JP",
      "name": "[Oniilus Refreshing Summer [English]",
      "tags": Array [
        19864,
        12227,
        2937,
        29859,
        16228,
        33172,
        6346,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720934/thumb.jpg",
    },
    Object {
      "id": 326868,
      "lang": "JP",
      "name": "(Cinderella☆Stage 8 STEP) [azure polka(Seichi-san)] azure idol graphics2 Airi Totoki (THE IDOLM@STER CINDERELLA GIRLS) [Digital]",
      "tags": Array [
        20925,
        7765,
        2937,
        17801,
        27378,
        20905,
        33172,
        6346,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720933/thumb.jpg",
    },
    Object {
      "id": 326867,
      "lang": "CN",
      "name": "[Sugota] Gyakuten School Caste [Chinese] [脸肿汉化组]",
      "tags": Array [
        29963,
        17249,
        35763,
        2515,
        2937,
        13720,
        85288,
        35762,
        20035,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720926/thumb.jpg",
    },
    Object {
      "id": 326866,
      "lang": "CN",
      "name": "(C96) [Rojiura Manhole (MAKI)] Roujima Mako no Ecchi na Aikata Boshuuchuu! [Chinese] [脸肿汉化组]",
      "tags": Array [
        29963,
        17249,
        90671,
        51068,
        13998,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720921/thumb.jpg",
    },
    Object {
      "id": 326865,
      "lang": "JP",
      "name": "(Utahime Teien 20) [Nekousa Pudding (Ra-men)] Munakata-san wa Ai saretai. (THE IDOLM@STER CINDERELLA GIRLS) [Digital]",
      "tags": Array [
        20925,
        28484,
        16643,
        21476,
        9388,
        35763,
        20525,
        24102,
        35762,
        33172,
        6346,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720923/thumb.jpg",
    },
    Object {
      "id": 326864,
      "lang": "CN",
      "name": "[Rocomani (Manatsu Roco)] DARIINA REVENGE!! (THE IDOLM@STER CINDERELLA GIRLS) [Chinese] [脸肿汉化组] [Digital]",
      "tags": Array [
        29963,
        17249,
        20925,
        16643,
        17871,
        74183,
        74184,
        10314,
        35762,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720920/thumb.jpg",
    },
    Object {
      "id": 326863,
      "lang": "CN",
      "name": "[Koyubi] Fujunai Process | 不纯爱Process Ch. 1-2 [Chinese] [拾荒者汉化组] [Digital]",
      "tags": Array [
        104201,
        29963,
        17249,
        85787,
        8378,
        21712,
        23895,
        8368,
        33173,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720910/thumb.jpg",
    },
    Object {
      "id": 326862,
      "lang": "JP",
      "name": "(C80) [Pale Pink! (Sakurabe Notos)] OP no kashi no imi wa yoku wakan'naikedo Kotori-chan wa kawaī (Yu-Gi-Oh! ZEXAL)",
      "tags": Array [
        26336,
        24127,
        25107,
        1624,
        17967,
        35763,
        19440,
        35762,
        33172,
        6346,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1713850/thumb.jpg",
    },
    Object {
      "id": 326861,
      "lang": "CN",
      "name": "[Tsukumo Nikyu] Hakata | Shape of teeth (COMIC Anthurium 2020-02) [Chinese] [贫穷恶魔个人汉化] [Digital]",
      "tags": Array [
        29963,
        17249,
        105605,
        35763,
        24380,
        35762,
        29366,
        33173,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720906/thumb.jpg",
    },
    Object {
      "id": 326860,
      "lang": "JP",
      "name": "[Yajirushi Key (Meito)] Noroi no Yubiwa de Game Over Kioku Kaihen Shuu",
      "tags": Array [
        90671,
        97545,
        22981,
        14283,
        35762,
        24201,
        31775,
        20905,
        27473,
        33172,
        6346,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720194/thumb.jpg",
    },
    Object {
      "id": 326859,
      "lang": "CN",
      "name": "[Hirune (Various)] Chinkasu x Hamigaki Fella x SeGero x Hana kara Semen Goudou[chinese]",
      "tags": Array [
        29963,
        17249,
        90671,
        122258,
        35163,
        112702,
        90403,
        33318,
        27720,
        25822,
        29859,
        15225,
        25766,
        108118,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720902/thumb.jpg",
    },
    Object {
      "id": 326858,
      "lang": "CN",
      "name": "[Inukami] Can't wait (COMIC Koh 2018-06) [Chinese] [黑条汉化] [Digital]",
      "tags": Array [
        29963,
        17249,
        78363,
        115879,
        24380,
        10314,
        33173,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1718386/thumb.jpg",
    },
    Object {
      "id": 326857,
      "lang": "CN",
      "name": "(COMIC1☆15) [Felt-Kobo (Flugel)] Maid na Wa-chan (Girls' Frontline) [Chinese] [v.v.t.m汉化组]",
      "tags": Array [
        29963,
        17249,
        74788,
        92923,
        88521,
        108416,
        108417,
        35763,
        29859,
        190,
        13720,
        35762,
        24201,
        8368,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720882/thumb.jpg",
    },
    Object {
      "id": 326856,
      "lang": "JP",
      "name": "[トロたく工場 (ko, Hypertrampoline, NOSTALISH)] Doragontakuse~tsu￮ su",
      "tags": Array [
        33658,
        70119,
        33903,
        75336,
        90492,
        14283,
        109360,
        28778,
        21712,
        85288,
        29023,
        23895,
        33172,
        6346,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1714692/thumb.jpg",
    },
    Object {
      "id": 326855,
      "lang": "EN",
      "name": "[Terasu MC] Kokujin no Tenkousei ni Haha o NTR ru [English][Bleached]",
      "tags": Array [
        12227,
        33252,
        17249,
        90671,
        104581,
        12824,
        30473,
        35763,
        2937,
        29224,
        2633,
        1207,
        13720,
        8653,
        85288,
        1590,
        6817,
        25794,
        20035,
        21572,
        104237,
        33172,
      ],
      "thumbnail": "https://t.nhentai.net/galleries/1720870/thumb.jpg",
    },
  ],
  12859,
]
`;
